### PR TITLE
Native coref component

### PR DESCRIPTION
**Pull request by svlandeg**
Work-in-progress

## Description
Creating a native coref component in spaCy

### Types of change
enhancement

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by svlandeg** [comment]
Status March 1:

- Wrote preliminary v3-compatible framework to facilitate experimentation with different coref models
- Currently assuming two different pipeline components:
    - `coref_er` is a rule-based mention detection algorithm: uses noun chunks, POS tags and named entities
    - `coref` assembles the provided mentions into clusters (dummy implementation)
- Using `doc.spans` to store the information:
    - `doc.spans[coref_mentions]` for storing all relevant coref mentions (nouns, pronouns, names, ...)
    - `doc.spans[coref_clusters_i]` for different clusters, indexed with `i` 
- `Coref.v0` needs to be implemented and changed to `Coref.v1`
- `Scorer.score_clusters` method that currently uses a too simple scoring mechanism (binary relations between mentions), should be refined with actual coref scoring algorithm

While all of this is mostly dummy framework, it already helped discover some bugs & required functionality, cf PRs https://github.com/explosion/spaCy/pull/7197, https://github.com/explosion/spaCy/pull/7209 and https://github.com/explosion/spaCy/pull/7225.

Going forward, having this bare framework should facilitate working on this functionality with different people in parallel, filling in different parts...

### TODO

- [ ] Implement proper `coref` ML model
- [ ] Proper mention detection algorithm, rule-based, ML-based, something like the [`SpanCategorizer`](https://github.com/explosion/spaCy/pull/6747), ...
- [ ] Meaningful evaluation script
- [ ] Tune & benchmark
- [ ] Rewrite errors to use `spacy.errors`

### Open questions / current issues

- While we [talked about](https://github.com/explosion/spaCy/pull/6696) keeping `doc.spans` a relatively simple dictionary of strings mapping to lists of spans, we might consider having a more formal way of defining clusters that belong together - currently this is done by matching [a prefix](https://github.com/explosion/spaCy/blob/5d0ec5ace48ab07c41f2aff1323a003dd1ef16c7/spacy/pipeline/coref.py#L106) in the `spans` key, which is obviously not ideal

- The design with the rule-based `coref_er` is again awkward, because this component won't run during `nlp.update`, meaning that the `coref` model could only train on gold mentions, which is not a good idea in terms of generalizability and robustness of the ML model.

File: spacy/ml/models/coref.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from typing import List
+from thinc.api import Model
+from thinc.types import Floats2d
+
+from ...util import registry
+from ...tokens import Doc
+
+
+@registry.architectures.register("spacy.Coref.v0")
```
Comment (honnibal) [inline spacy/ml/models/coref.py:9]: I think we normally use this as a callable, so just `registry.architectures`?

Comment (svlandeg) [reply spacy/ml/models/coref.py:9]: I'm not sure I understand what you mean?

Comment (svlandeg) [reply spacy/ml/models/coref.py:9]: Oh wait, you're right, like we'd do `@registry.loggers`. For some reason the architectures always use `.register` - I can clean that up on the `master` branch.

**Comment by svlandeg** [comment]
I'll just merge this into the new WIP `feature/coref` branch on explosion, so we can continue the discussion there.

**Pull request merged**